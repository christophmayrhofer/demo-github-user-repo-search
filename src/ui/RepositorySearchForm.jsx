import React, { PropTypes } from 'react';
import Immutable from 'immutable';
import ImmutablePropTypes from 'react-immutable-proptypes';
import { connect } from 'react-redux';
import Spinner from 'react-spin';
import RepositoryList from './RepositoryList';
import { updateUsername, fetchUserRepositories } from '../redux/actions';

export const RepositorySearchFormComponent = ({
    username = 'christophmayrhofer',
    hasNeverFetched = true,
    isFetching = false,
    hasFetchError = false,
    errorMessage = '',
    repositories = Immutable.List.of(),
    onChangeUsername = () => {},
    onClickFetch = () => {},
  }) => (
    <div className="flex">
      <form className="form-inline full-width">
        <input id="username" type="text" className="form-control" placeholder="Github Username" value={username} onChange={onChangeUsername} />
        <input id="submit" type="submit" className="btn btn-primary" onClick={onClickFetch} value="Fetch Repositories" />
      </form>

      <Choose>
        <When condition={hasNeverFetched}>
          <div className="alert alert-info">
            Enter Github username and click fetch...
          </div>
        </When>
        <When condition={isFetching}>
          <Spinner />
        </When>
        <When condition={hasFetchError}>
          <div className="alert alert-danger">
            {errorMessage}
          </div>
        </When>
        <Otherwise>
          <RepositoryList repositories={repositories} />
        </Otherwise>
      </Choose>
    </div>
);

RepositorySearchFormComponent.propTypes = {
  username: PropTypes.string.isRequired,
  hasNeverFetched: PropTypes.bool.isRequired,
  isFetching: PropTypes.bool.isRequired,
  hasFetchError: PropTypes.bool.isRequired,
  errorMessage: PropTypes.string,
  repositories: ImmutablePropTypes.listOf(
    ImmutablePropTypes.contains({
      id: PropTypes.number.isRequired,
      name: PropTypes.string.isRequired,
      html_url: PropTypes.string.isRequired,
    })
  ),
  onChangeUsername: PropTypes.func.isRequired,
  onClickFetch: PropTypes.func.isRequired,
};

const mapStateToProps = state => ({
  username: state.getIn(['repositoriesSearch', 'username']),
  hasNeverFetched: state.getIn(['repositoriesSearch', 'hasNeverFetched']),
  isFetching: state.getIn(['repositoriesSearch', 'isFetching']),
  hasFetchError: state.getIn(['repositoriesSearch', 'hasFetchError']),
  errorMessage: state.getIn(['repositoriesSearch', 'errorMessage']),
  repositories: state.getIn(['repositoriesSearch', 'repositories']),
});
const mapDispatchToProps = dispatch => ({
  onClickFetch: (event) => { event.preventDefault(); dispatch(fetchUserRepositories()); },
  onChangeUsername: event => dispatch(updateUsername(event.target.value)),
});

const RepositorySearchForm = connect(
  mapStateToProps,
  mapDispatchToProps
)(RepositorySearchFormComponent);

export default RepositorySearchForm;
