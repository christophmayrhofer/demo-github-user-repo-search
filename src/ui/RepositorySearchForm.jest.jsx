import React from 'react';
import Immutable from 'immutable';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import { RepositorySearchFormComponent } from './RepositorySearchForm';

describe('RepositorySearchForm', () => {
  const baseState = Immutable.fromJS({
    username: 'username',
    hasNeverFetched: false,
    isFetching: false,
    hasFetchError: false,
    errorMessage: '',
    repositories: Immutable.List.of(),
    onChangeUsername: jest.fn(),
    onClickFetch: jest.fn(),
  });

  it('renders hasNeverFetched message if the fetch button has never been pressed', () => {
    const props = baseState.merge({ hasNeverFetched: true }).toObject();

    const wrapper = shallow(
      <RepositorySearchFormComponent {...props} />
    );

    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  it('renders a spinner if isFetching is true', () => {
    const props = baseState.merge({ hasNeverFetched: false, isFetching: true }).toObject();

    const wrapper = shallow(
      <RepositorySearchFormComponent {...props} />
    );

    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  it('renders errorMessage if hasFetchError is true', () => {
    const props = baseState.merge({
      hasNeverFetched: false,
      isFetching: false,
      hasFetchError: true,
      errorMessage: 'errorMessage',
    }).toObject();

    const wrapper = shallow(
      <RepositorySearchFormComponent {...props} />
    );

    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  it('renders repositoryList after fetching', () => {
    const props = baseState.merge({
      repositories: Immutable.List.of(),
    }).toObject();

    const wrapper = shallow(
      <RepositorySearchFormComponent {...props} />
    );

    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  it('calls onChangeUsername when typing in the username input field', () => {
    const props = baseState.merge({
      onChangeUsername: jest.fn(),
    }).toObject();

    const wrapper = shallow(
      <RepositorySearchFormComponent {...props} />
    );

    wrapper.find('#username').simulate('change', 'username');
    expect(props.onChangeUsername).toBeCalledWith('username');
  });

  it('calls onClickFetch when the fetch button is clicked', () => {
    const props = baseState.merge({
      onClickFetch: jest.fn(),
    }).toObject();

    const wrapper = shallow(
      <RepositorySearchFormComponent {...props} />
    );

    wrapper.find('#submit').simulate('click');
    expect(props.onClickFetch).toBeCalled();
  });
});
