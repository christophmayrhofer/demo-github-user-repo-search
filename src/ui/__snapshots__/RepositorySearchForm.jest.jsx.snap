exports[`RepositorySearchForm renders a spinner if isFetching is true 1`] = `
<div
  className="flex">
  <form
    className="form-inline full-width">
    <input
      className="form-control"
      id="username"
      onChange={[Function]}
      placeholder="Github Username"
      type="text"
      value="username" />
    <input
      className="btn btn-primary"
      id="submit"
      onClick={[Function]}
      type="submit"
      value="Fetch Repositories" />
  </form>
  <ReactSpinner />
</div>
`;

exports[`RepositorySearchForm renders errorMessage if hasFetchError is true 1`] = `
<div
  className="flex">
  <form
    className="form-inline full-width">
    <input
      className="form-control"
      id="username"
      onChange={[Function]}
      placeholder="Github Username"
      type="text"
      value="username" />
    <input
      className="btn btn-primary"
      id="submit"
      onClick={[Function]}
      type="submit"
      value="Fetch Repositories" />
  </form>
  <div
    className="alert alert-danger">
    errorMessage
  </div>
</div>
`;

exports[`RepositorySearchForm renders hasNeverFetched message if the fetch button has never been pressed 1`] = `
<div
  className="flex">
  <form
    className="form-inline full-width">
    <input
      className="form-control"
      id="username"
      onChange={[Function]}
      placeholder="Github Username"
      type="text"
      value="username" />
    <input
      className="btn btn-primary"
      id="submit"
      onClick={[Function]}
      type="submit"
      value="Fetch Repositories" />
  </form>
  <div
    className="alert alert-info">
    Enter Github username and click fetch...
  </div>
</div>
`;

exports[`RepositorySearchForm renders repositoryList after fetching 1`] = `
<div
  className="flex">
  <form
    className="form-inline full-width">
    <input
      className="form-control"
      id="username"
      onChange={[Function]}
      placeholder="Github Username"
      type="text"
      value="username" />
    <input
      className="btn btn-primary"
      id="submit"
      onClick={[Function]}
      type="submit"
      value="Fetch Repositories" />
  </form>
  <RepositoryList
    repositories={Array []} />
</div>
`;
